urlfd018a75f7_highlanderkev_StockExVirtuoso_tgz-pJ8-driver_Phase4J8/src/book/ProductBook.java

```diff
@@ -0,0 +1,457 @@
+    }
+
+    /**
+     * This method will cancel the specified user's quote on both the buy and
+     * sell sides.
+     *
+     * @param userName of user
+     * @throws Exception
+     */
+    public synchronized void cancelQuote(String userName) throws Exception {
+        buyBook.submitQuoteCancel(userName);
+        sellBook.submitQuoteCancel(userName);
+        updateCurrentMarket();
+    }
+
+    /**
+     * This method will add the provided quote side to the Buy and Sell product
+     * book sides.
+     *
+     * @param q
+     * @throws DataValidationException
+     * @throws Exception
+     */
+    public synchronized void addToBook(Quote q) throws DataValidationException, Exception {
+        Price sellQuote = q.getQuoteSide(BookSide.SELL).getPrice();
+        Price buyQuote = q.getQuoteSide(BookSide.BUY).getPrice();
+        sellQuote = check(sellQuote);
+        buyQuote = check(buyQuote);
+        int sellVolume = q.getQuoteSide(BookSide.SELL).getOriginalVolume();
+        int buyVolume = q.getQuoteSide(BookSide.BUY).getOriginalVolume();
+        if (sellQuote.lessOrEqual(buyQuote)) {
+            throw new DataValidationException("DataValidationException: Sell price is less than or equal to buy price.");
+        } else if (buyQuote.lessOrEqual(PriceFactory.makeLimitPrice("00")) || sellQuote.lessOrEqual(PriceFactory.makeLimitPrice("00"))) {
+            throw new DataValidationException("DataValidationException: Price is less than or equal to zero.");
+        } else if (buyVolume <= 0 || sellVolume <= 0) {
+            throw new DataValidationException("DataValidationException: Volume is less than or equal to zero.");
+        } else {
+            String userName = q.getUserName();
+            if (userQuotes.contains(userName)) {
+                buyBook.removeQuote(userName);
+                sellBook.removeQuote(userName);
+            }
+            if (ExceptionHandler.checkObject(q, "book.ProductBook#addtoBook.")) {
+                Tradable buyQ = q.getQuoteSide(BookSide.BUY);
+                Tradable sellQ = q.getQuoteSide(BookSide.SELL);
+                addToBook(BookSide.BUY, buyQ);
+                addToBook(BookSide.SELL, sellQ);
+                userQuotes.add(userName);
+                updateCurrentMarket();
+            }
+        }
+    }
+
+    /**
+     * This method will add the provided order to the appropriate product book
+     * side.
+     *
+     * @param o order
+     * @throws Exception
+     */
+    public synchronized void addToBook(Order o) throws Exception {
+        addToBook(o.getSide(), o);
+        updateCurrentMarket();
+    }
+
+    /**
+     * This method is a key part of the trading system and deals with the
+     * addition of Tradable objects to the BUY/SELL product book side, handling
+     * the results of any trades.
+     *
+     * @param side of book
+     * @param t tradable
+     */
+    private synchronized void addToBook(BookSide side, Tradable t) throws Exception {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 12. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.